lib/PrintAsObjC/PrintAsObjC.cpp

```diff
@@ -479,7 +487,9 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
 
     // Swift designated initializers are Objective-C designated initializers.
     if (auto ctor = dyn_cast<ConstructorDecl>(AFD)) {
-      if (ctor->isDesignatedInit() &&
+      if (ctor->hasStubImplementation() || ctor->getFormalAccess() < minRequiredAccess) {
```

Comment:
Ok. While I'm at it I'm going to tweak the tests as well to test the grandchild-reintroduces-initializer case.